src/test/codegen/issue-73338-effecient-cmp.rs

```diff
@@ -0,0 +1,39 @@
+// This test checks that comparison operation
+// generated by #[derive(PartialOrd)]
+// doesn't contain jumps for C enums
+
+// compile-flags: -Copt-level=3
+
+#![crate_type="lib"]
+
+#[repr(u32)]
+#[derive(Copy, Clone, Eq, PartialEq, PartialOrd)]
+pub enum Foo {
+    Zero,
+    One,
+    Two,
+}
+
+#[no_mangle]
+pub fn compare_less(a: Foo, b: Foo)->bool{
+    // CHECK-NOT: br {{.*}}
+    a < b
+}
+
+#[no_mangle]
+pub fn compare_le(a: Foo, b: Foo)->bool{
+    // CHECK-NOT: br {{.*}}
+    a <= b
+}
+
+#[no_mangle]
+pub fn compare_ge(a: Foo, b: Foo)->bool{
+    // CHECK-NOT: br {{.*}}
+    a >= b
+}
+
+#[no_mangle]
+pub fn compare_greater(a: Foo, b: Foo)->bool{
+    // CHECK-NOT: br {{.*}}
+    a > b
```

Comment:
It's interesting that this did not need a modification to ge -- maybe hints at some LLVM bug?

Reply:
Well, the `Option<Ordering>` has such u8 values 
```
Some(Less) == 255
Some(Equal) == 0
Some(Greater) == 1
None == 2
```
so, `>=` compiles to `cmp_result as u8 < 2` which easy to optimize.

Old implementation of `<=` couldn't be optimized such easily because it was `(255 == cmp_result || 0 == cmp_result)` which LLVM failed to optimize. After my change it becomes `!(cmp_result > 0)` which optimized much better.

Reply:
Possibly, current version of LLVM handle comparison with 2 consequtive numbers better than with 2 different numbers.

Reply:
Makes sense, OK. Thanks!

Reply:
Any idea what the fallout would be if `Ordering` was changed from `(Less, Equal, Greater) == (-1, 0, 1)` to `(Less, Equal, Greater) == (0, 1, 2)`? I didn't find a guarantee for the underlying values being stable and `impl Ord for Ordering` only relies on `Less < Equal < Greater`.

Reply:
I don't know enough for this, I think.

Current solution could be profitable in conversion from `memcmp` results (compiler can do `sign(memcmp(a, b))` to get Ordering).